drivers/ieee802154/ieee802154_rf2xx_iface.c

```diff
@@ -0,0 +1,312 @@
+		{
+			.buf = &addr,
+			.len = 1
+		},
+		{
+			.buf = &data,
+			.len = 1
+		}
+	};
+	const struct spi_buf_set tx = {
+		.buffers = tx_buf,
+		.count = 2
+	};
+	const struct spi_buf rx_buf = {
+		.buf = &status,
+		.len = 1
+	};
+	const struct spi_buf_set rx = {
+		.buffers = &rx_buf,
+		.count = 1
+	};
+	if (spi_transceive(ctx->spi, &ctx->spi_cfg, &tx, &rx) != 0) {
+		LOG_ERR("Failed to exec rf2xx_reg_write at address %d",
+			addr);
+	}
+
+	LOG_DBG("Write Address: %02X, PhyStatus: %02X, RegVal: %02X",
+		(addr & ~(RF2XX_RF_CMD_REG_W)), status, data);
+}
+
+u8_t rf2xx_iface_bit_read(struct device *dev,
+			  u8_t addr,
+			  u8_t mask,
+			  u8_t pos)
+{
+	u8_t ret;
+
+	ret = rf2xx_iface_reg_read(dev, addr);
+	ret &= mask;
+	ret >>= pos;
+
+	return ret;
+}
+
+void rf2xx_iface_bit_write(struct device *dev,
+			   u8_t reg_addr,
+			   u8_t mask,
+			   u8_t pos,
+			   u8_t new_value)
+{
+	u8_t current_reg_value;
+
+	current_reg_value = rf2xx_iface_reg_read(dev, reg_addr);
+	current_reg_value &= ~mask;
+	new_value <<= pos;
+	new_value &= mask;
+	new_value |= current_reg_value;
+	rf2xx_iface_reg_write(dev, reg_addr, new_value);
+}
+
+void rf2xx_iface_frame_phr(struct device *dev,
+			   u8_t *phr)
+{
+	struct rf2xx_context *ctx = dev->driver_data;
+	u8_t cmd = RF2XX_RF_CMD_FRAME_R;
+	u8_t status;
+
+	const struct spi_buf tx_buf = {
+		.buf = &cmd,
+		.len = 1
+	};
+	const struct spi_buf_set tx = {
+		.buffers = &tx_buf,
+		.count = 1
+	};
+	const struct spi_buf rx_buf[2] = {
+		{
+			.buf = &status,
+			.len = 1
+		},
+		{
+			.buf = phr,
+			.len = 1
+		},
+	};
+	const struct spi_buf_set rx = {
+		.buffers = rx_buf,
+		.count = 2
+	};
+
+	if (spi_transceive(ctx->spi, &ctx->spi_cfg, &tx, &rx) != 0) {
+		LOG_ERR("Failed to exec rf2xx_frame_read PHR");
+	}
+
+	LOG_DBG("Read PHR: PhyStatus: %02X, PHR: %02X",
+		status, *phr);
+}
+
+void rf2xx_iface_frame_read(struct device *dev,
+			    u8_t *data,
+			    u8_t length,
+			    u8_t *lqi,
+			    u8_t *ed,
+			    enum rf2xx_trx_state_trac_t *trac)
+{
+	struct rf2xx_context *ctx = dev->driver_data;
+	u8_t cmd = RF2XX_RF_CMD_FRAME_R;
+	u8_t status;
+	u8_t phr;
+
+	const struct spi_buf tx_buf = {
+		.buf = &cmd,
+		.len = 1
+	};
+	const struct spi_buf_set tx = {
+		.buffers = &tx_buf,
+		.count = 1
+	};
+	const struct spi_buf rx_buf[6] = {
+		{
+			.buf = &status,
+			.len = 1
+		},
+		{
+			.buf = &phr,
+			.len = 1
+		},
+		{
+			.buf = data,
+			.len = length
+		},
+		{
+			.buf = lqi,
+			.len = 1
+		},
+		{
+			.buf = ed,
+			.len = 1
+		},
+		{
+			.buf = trac,
+			.len = 1
+		},
+	};
+	const struct spi_buf_set rx = {
+		.buffers = rx_buf,
+		.count = 6
+	};
+
+	if (spi_transceive(ctx->spi, &ctx->spi_cfg, &tx, &rx) != 0) {
+		LOG_ERR("Failed to exec rf2xx_frame_read");
+	}
+
+	*trac = (*trac >> RF2XX_TRAC_STATUS) & 7;
+
+	LOG_DBG("Read Frame: PhyStatus: %02X, PHR: %02X, LQI: %02X, ED: %02X,"
+		"TRAC: %02X", status, phr, *lqi, *ed, *trac);
+}
+
+void rf2xx_iface_frame_write(struct device *dev,
```

Comment:
same from frame_phr, frame_read and frame_write. There seems to be rooms for factorization.

Reply:
This case is worst than reg_read/write because I need read something by SPI that I don't know the length (two transactions) or execute one transaction with max size (memory issues). So, I did a test and merged frame_read and frame_phr but still need to use 2 transactions. This merge need 132 bytes on stack at rf2xx_trx_rx. In case of merge read/write the spi buffers are to different and will be more difficult the read code. It seams to me that maintain the method with one specific purpose will at end be better.

Reply:
I merged `frame_read` and `frame_phr` and improved `rf2xx_trx_rx`. For now I'll keep `frame_read` and `frame_write` separated.